### PR TITLE
Shell

### PR DESCRIPTION
**Pull request by hughleat**
Some changes to the manual_env.

Allows command completion. 
Adds an undoable stack of actions.
Adds a couple of simple search techniques.
Adds a few miscellaneous capabilities.

There are some issues that are going to be or have been fixed in the mainline.
1. After downloading a new dataset the benchmarks do not update without restarting
2. The benchmark is not set from the command line

### CODE REVIEW

**Comment by hughleat** [comment]
Oh, one more thing. When I set the default reward or observation I get this:

/Users/hleather/opt/miniconda3/lib/python3.8/site-packages/compiler_gym-0.1.2-py3.8.egg/compiler_gym/envs/compiler_env.py:503: UserWarning: Changing eager observation space has no effect until reset() is called.
  warnings.warn(
Observation  in 3.7ms

Not sure if that warning is necessary or if there is better way for me to use it.

File: compiler_gym/bin/manual_env.py
Comment (ChrisCummins) [inline compiler_gym/bin/manual_env.py:585]: Could we add a `greedy -` that runs greedy until no action is selected?

File: compiler_gym/bin/manual_env.py
Comment (ChrisCummins) [inline compiler_gym/bin/manual_env.py]: If `i != num_steps` it would be nice to print a message to explain why. Perhaps something like "Greedy search for 10 steps stopped after step 5 selected no action"?

File: compiler_gym/bin/manual_env.py
Comment (ChrisCummins) [inline compiler_gym/bin/manual_env.py]: FYI (no action needed) there is a shortcut for this:

```suggestion
        with Timer("Got actions") as timer:
            items = self.get_action_rewards()
```

File: compiler_gym/bin/manual_env.py
Comment (ChrisCummins) [inline compiler_gym/bin/manual_env.py]: nitpick: I'd just call these `observation`/`reward` to keep in line with regular gym terminology. We'll drop the "eager" branding soon

same applies for other uses of `eager_` in this file

File: compiler_gym/bin/manual_env.py
Comment (ChrisCummins) [inline compiler_gym/bin/manual_env.py]: Consider calling `reward` with no arg to print the default reward, if set. At the moment it errors:
```
compilergym:cBench-v0/dijkstra> reward
Unknown reward, ''
Rewards are listed with command, list_rewards
```

instead:

```
compilergym:cBench-v0/dijkstra> reward
0.000000
Reward IrInstructionCount in 667.8us
```

Same for `observation` with no arg

File: compiler_gym/bin/manual_env.py
Comment (ChrisCummins) [inline compiler_gym/bin/manual_env.py]: Not urgent, but it would be nice if hill climbing printed a summary at the end, something like:

`Hill climb completed in 2.12s. Accepted 4 of 100 steps for a total reward of 0.123`

**Review by ChrisCummins — CHANGES_REQUESTED** [review]
This is great, thank you! I love the tab completion and the new stack/search features. Some minor code nitpicks inline.

My biggest request is for more documentation. The old [manual_env](https://facebookresearch.github.io/CompilerGym/cli.html#module-compiler_gym.bin.manual_env) doesn't have much to say as it can't do much, but your new version is useful enough that it deserves a much more verbose description, maybe even a mini-tutorial. Could you elaborate the module docstring to add something like this?

A welcome message on startup that points the user towards `help` may also be good, presently it drops the user into a commandline without much fanfare:

```
$ python -m compiler_gym.bin.manual_env --env=llvm-ic-v0
Initialized environment in 125.7ms
compilergym:NO-BENCHMARK>
```

BTW, since I've never used the cmd.Cmd class, is it easy to test? Could we have a smoke test that simulates the user selecting some known benchmark and running say a single action?

Cheers,
Chris

File: compiler_gym/bin/manual_env.py
Comment (hughleat) [reply compiler_gym/bin/manual_env.py]: Done.

File: compiler_gym/bin/manual_env.py
Comment (hughleat) [reply compiler_gym/bin/manual_env.py]: Good one. I've changed those that can be changed.

File: compiler_gym/bin/manual_env.py
Comment (hughleat) [reply compiler_gym/bin/manual_env.py]: Cool. Wasn't sure if the others had changed.

File: compiler_gym/bin/manual_env.py
Comment (hughleat) [reply compiler_gym/bin/manual_env.py]: Done

File: compiler_gym/bin/manual_env.py
Comment (hughleat) [reply compiler_gym/bin/manual_env.py]: Done

File: compiler_gym/bin/manual_env.py
Comment (hughleat) [reply compiler_gym/bin/manual_env.py:585]: Gonna leave that one for the moment. Can we leave it to a feature req (I'm a bit desperate to get back to the deep DF stuff).

Comment (ChrisCummins) [reply compiler_gym/bin/manual_env.py:585]: Yep sounds good!

Comment (ChrisCummins) [reply compiler_gym/bin/manual_env.py:585]: Follow up in #75

File: compiler_gym/bin/manual_env.py
Comment (ChrisCummins) [reply compiler_gym/bin/manual_env.py]: Thanks!

**Comment by ChrisCummins** [comment]
Great, thanks for the changes @hughleat! Love the tests :)

One tiny thing:
```
compilergym:NO-BENCHMARK> tutorial
*** Unknown syntax: tutorial
```

I don't think `tutorial` works. In fact, could you just move the tutorial text into the module docstring? That way it'll get picked up and rendered in the reference docs: https://facebookresearch.github.io/CompilerGym/cli.html#module-compiler_gym.bin.manual_env

**Comment by ChrisCummins** [comment]
BTW the CI tests were failing. Not sure why, as I can reproduce locally just fine. I've added a small patch to make it easier to debug 8d642118e1bdfcb6449a2a9133df1bbfa8686e65

**Comment by hughleat** [comment]
> I don't think `tutorial` works. In fact, could you just move the tutorial text into the module docstring? That way it'll get picked up and rendered in the reference docs: https://facebookresearch.github.io/CompilerGym/cli.html#module-compiler_gym.bin.manual_env

Changed it from `help tutorial` to `tutorial`. Now extracts from `__doc__`.

**Review by ChrisCummins — APPROVED** [review]
Thanks for the updates @hughleat! I added a small fix for a test-execution-order bug 7e88bc763bc39bdf070ed88f6bbf2c80045dbcff.

LGTM, please merge when you're happy :)

**Pull request merged**
